### PR TITLE
Allow creating custom block templates in classic themes

### PR DESCRIPTION
**Pull request by youknowriad**
The idea of this PR is to allow creating custom block templates (starting blank) for a given page in classic themes. It's a way for classic themes to enable some FSE features without any change.

**How to test it**

It's very rough at the moment, more a POC, but here's how  you do  it:

 - Create  a new page
 - Save it
 - you'll notice on the Document sidebar, a "Create custom template" link
 - Once you click that link a custom block template is created and attached to the page.
 - By default that template Is composed by just two blocks (post title and post content) but you can edit as you wish
 - Apply the changes and publish

Now, on the frontend, your page is using its own custom template and not the regular `page.php` or `singular.php` that comes from the theme itself.

**Notes**

This PR can be seen as a migration strategy for classic themes, it highlights a few things in terms of code/architecture:

 - The is_fse_theme checks become less important as this blurs the line between classic and FSE themes, the whole FSE infrastructure (template post type, template resolution, template API) need to be available for classic themes too.
 - In terms of design it's very rough at the moment, we need to improve that. I'm looking at you designers :P @jasmussen @jameskoster 
 - All FSE blocks become available for classic themes with this PR, this means if we ship this to Core, we need to make some good decisions about which FSE blocks are considered stable.
 - Right  now, it's enable for all classic themes by default, we might want to introduce a theme support flag for it to make it opt-in.

**Todo**

 - [x] Allow setting titles when creating custom templates.
 - [x] Improve the design a bit.
 - [x] Write an e2e test.
 - [x] Add some documentation.
 - [x] Theme support flag.

### CODE REVIEW

**Comment by youknowriad** [comment]
This PR is not ready yet but it's ready to get thoughts and designs.

**Comment by aristath** [comment]
> Right now, it's enable for all classic themes by default, we might want to introduce a theme support flag for it to make it opt-in.

I think that would be necessary in many situations... and it should also be filterable so users can install a plugin to enable it on their old, not-updated theme. See PR https://github.com/WordPress/gutenberg/pull/30465 for a suggestion

**Comment by aristath** [comment]
Thinking about the theme-support thing some more, perhaps it should be opt-out instead of opt-in? 🤔 
IMO this is going to be one of the things that will get huge adoption by classic-theme users... It will allow them to use their theme for their site, while at the same time being able to build a completely custom frontpage if they so desire.
If we make it opt-in we'll be excluding a large portion of the userbase, but it is the "safer" choice because people just won't see the feature.
If we make it opt-out, then we are more inline with what we want the future of WP to be. If a theme doesn't want FSE for some reason, then they can explicitly define that.

I believe opt-out would make more sense... (updated the PR on #30465 to reflect that)

File: packages/edit-post/src/store/actions.js [hunk 1]
```diff
@@ -436,3 +440,47 @@ export function setIsEditingTemplate( value ) {
...
+export function* __unstableSwitchToEditingMode( shouldCreateTemplate = false ) {
+	if ( shouldCreateTemplate ) {
+		const templateContent = [
+			createBlock( 'core/post-title' ),
+			createBlock( 'core/post-content' ),
+		];
+		const template = {
+			slug: 'custom-template-' + uuid(),
+			content: serialize( templateContent ),
+			title: 'Custom Template',
```
Comment (youknowriad) [inline packages/edit-post/src/store/actions.js]: To avoid having all custom templates have the same name by default, I think we should do something similar to template parts and reusable block: show a modal when clicking the "Create custom template" link and ask the user to fill a title.

Comment (aristath) [reply packages/edit-post/src/store/actions.js]: A title would be useful in identifying the templates 👍

**Comment by paaljoachim** [comment]
What an intriguing PR!

I am using the Twenty Twenty One theme.
I downloaded a special version of the Gutenberg plugin (Going to Checks -> clicked Gutenberg build in left sidebar -> downloaded Gutenberg plugin. Extracted and compressed it again and uploaded the special GB plugin. Additional instructions here: https://youtu.be/eJhgUjrTSF0). 

Then followed the instructions added by Riad.
Create a page. I saved it as a draft. I did not see a "Create custom template" link in the sidebar. How do I proceed?

**Comment by aristath** [comment]
@paaljoachim That was my bad, in https://github.com/WordPress/gutenberg/pull/30438/commits/6067b4236b28ca38ee7b92abc1c6322ad5d61270 the line defining theme-support was not renamed. I pushed a fix here just now so once the build finishes you'll be able to test it 👍 
If you don't want to wait for the build, you could add in your functions.php file `add_theme_support( 'block-templates' );`

**Comment by paaljoachim** [comment]
Great! In addition to using the special Gutenberg plugin mentioned above I then added the code snippet to the functions file `add_theme_support( 'block-templates' );` 

My test.

1.I would suggest a developer create a PR for moving the page/post title to the top bar. It would also be one of the many connections with Full Site Editing. https://github.com/WordPress/gutenberg/issues/27093

As having the title in the top bar. Be it "Page content: Page title", "Content: Page title" or just the "Page title". 
Would right away create a connection with using templates.

For this PR where I test page templates.
One should then see the page title in the top bar. Click the Create custom template" Then in the top bar see 
Template: Page title. Having the Template: Page title in the top bar gives a nice signal to the user where they are. 
Here is a mockup where I added the "**Custom template**: Test template" to the top bar and "Custom template: Document" to the bottom breadcrumbs. Having multiple ways for the user to see where they are. (In addition to seeing the word template and template name in the sidebar settings.)

Custom-template

2.Save or do not save is the question. Save or discard a save.
User clicks "Apply"  and notices the prechecked Custom Template. 

Screen Shot 2021-04-02 at 16 34 23


What if the user decides not to save?
They see this:

Screen Shot 2021-04-02 at 16 35 20

A greyed out blue button. User it not able to move onward, and meets a dead end.
An alternative suggestion would be as seen in this comment:
https://github.com/WordPress/gutenberg/issues/29577#issuecomment-812544687

Which basically does this:
Discard-custom-template-save

The user makes a selection if they want or do not want to save the custom template. Perhaps they just wanted to see what it is without saving. If they decide not to save they become stuck in the custom template. Having no obvious way on how to get back to the regular page. Perhaps they click the Page tab and look at Status & visibility to see if there is a bridge back to the page content in the same place they clicked to "Create a custom template" but there is none. So the user is forced to save. 

Screen Shot 2021-04-02 at 16 44 00

Going back to the page content. It looks like this...

Screen Shot 2021-04-02 at 16 44 41

User might think. Why am I not seeing the template content in my page? 
As it should show up similar to how a Reusable block is seen. Perhaps even use the click through method being explored here: https://github.com/WordPress/gutenberg/pull/30156 (Actually James @jameskoster  has continued to explore the click through method.) As the user would initially click the template and select the parent container. Then click into it and make changes. Then Save the change in some way. Either with a Save button in the toolbar as suggested by @critterverse Channing here: https://github.com/WordPress/gutenberg/issues/30357 or having a save button in the sidebar settings.

Custom-template-similar-to-Reusable-block

Bottom line is that being able to add a page template is a very nice feature! Along the way before this PR is merged other PR's can be created making specific UI (top bar page title) and UX (being able to easily discard a save) making the flow easier for the general user and also for us. Thanks!

**Comment by jasmussen** [comment]
Thanks for working on this one. One concern to keep in mind for the design depends on our ability to keep the user experience the same between themes and block themes (some of @jameskoster's FSE issues showcase where we might go with this).

That is, if we can keep the two flows mostly the same (except for the missing template parts like header and footer for non-block-themes) then this will mitigate having to relearn the flow when swapping to a block theme.

If we can't keep the two flows mostly the same, can you help me understand at which points the paths diverge the most? I took the branch for a quick spin, and the primary differentiator for me was the lack of a header and footer appearing in template mode (making the canvas size change @jameskoster and I have been back and forth on all the more urgent). Are there other parts of the flow?

And this one:

Screenshot 2021-04-06 at 10 09 17

Is this meant to be a way to create custom templates for block themes as well? Or is it a button created in absence of the Site Editor being available for non-block-themes?

Finally, I was able to create and save a new custom page template, where I inserted a template part and a post content part. But those were not to be found when I swapped out my page template on a subsequent page creation screen — did I miss a part of the flow here?

**Comment by youknowriad** [comment]
This is actually exactly the same thing as the template mode  in FSE themes. So flows should  be exactly the same.

The lack of header and footer is just the fact that we're starting "blank" (so not an existing template), I've just used a default template (title + content). In fact, the same flow can be useful for FSE themes as well to bootstrap new custom templates. 

I believe we might not want to ship "template parts" in the first version of this (if it ships sooner than FSE) but these can work without problems.

> But those were not to be found when I swapped out my page template on a subsequent page creation screen — did I miss a part of the flow here?

Not sure what you mean here.

---

Can you point me to explorations/designs about how to place the "create new template" link/button. For me that's the important thing here in this PR. The rest is just using the existing template mode designs without any difference, so any change affect both and is kind of separate from the current PR.

**Comment by jasmussen** [comment]
> This is actually exactly the same thing as the template mode in FSE themes. So flows should be exactly the same.

Happy to hear that, so presumably the improvements we make to the latter will benefit this as well?

> The lack of header and footer is just the fact that we're starting "blank" (so not an existing template), I've just used a default template (title + content). In fact, the same flow can be useful for FSE themes as well to bootstrap new custom templates.

👍 👍 it does still speak to needing a more stark visual context shift.

> I believe we might not want to ship "template parts" in the first version of this (if it ships sooner than FSE) but these can work without problems.

👍 👍

> Not sure what you mean here.

When I tested, I created a new page, called "Custom Template", then pressed the button in the sidebar, and added a new header template part that simply output "header", and a new post content loop below, to presumably hold the page contents (I'm fresh off block theme development here). 

And when I then created a new page from scratch, choosing the "Custom Template" that now showed up in the template dropdown, the "header" bit was not to be found when I published. 

> Can you point me to explorations/designs about how to place the "create new template" link/button. For me that's the important thing here in this PR. The rest is just using the existing template mode designs without any difference, so any change affect both and is kind of separate from the current PR.

[@jameskoster](https://github.com/WordPress/gutenberg/issues?q=is%3Aopen+is%3Aissue+label%3A%22%5BFeature%5D+Full+Site+Editing%22+author%3Ajameskoster) has been doing amazing work, but I'm having a hard time to pinpoint the exact place in the comments where we discussed this. But I think the solution lies somewhere between [this](https://github.com/WordPress/gutenberg/issues/27849#issuecomment-800994805) and [this](https://github.com/WordPress/gutenberg/issues/27849#issuecomment-800146730). It's just crucially important we get the design super right before we commit to either path, because as you can see it's a bit of a shift.

**Comment by jameskoster** [comment]
Afaik we haven't actually explored the creation flows for $custom templates yet, as that is not a feature we present in the FSE UI. Currently users are only able to create specific templates (page, search, 404, etc).

I agree careful consideration needs to be taken here, as once the custom template is created the "Template" section of the "Status & visibility" panel needs to accommodate more actions:

* Switch template
* Edit template
* Create new template (from scratch. IE the flow in this PR)
* Create new template (fork of existing template, $custom or otherwise)
* Create new template (page-$id)

I appreciate the last two may be some way out, but they should probably be included in design explorations.

It may be as simple as displaying a row of links, but I suspect something (a little) more elaborate may be required to better communicate the nuance of these actions.

---

I think the lack of Header in the new template may be a little confusing. Folks may expect their regular header to be included and displayed, but there's no clear indication as to whether this is (or isn't) the case. Maybe we should automatically spawn a basic Header template part behind the scenes on the users behalf as a part of this flow?

**Comment by youknowriad** [comment]
> Maybe we should automatically spawn a basic Header template part behind the scenes on the users behalf as a part of this flow?

What should this header template contain? There's no way to know what the current theme contains.

**Comment by youknowriad** [comment]
Capture d’écran 2021-04-06 à 10 13 29 AM

I tweaked the design a bit and enabled this flow for FSE themes as well now. I think we're now in a good position to ship a V1 of this PR and iterate on the designs of the template mode separately.

**Comment by youknowriad** [comment]
I'll try to write some documentation for this and add an e2e test a bit later.

**Comment by aristath** [comment]
For the header/footer template-parts in classic themes, this may be useful: https://github.com/WordPress/gutenberg/pull/30345
It would provide a way for classic themes to become hybrid by providing block templates for their header & footer.
So if a theme has these, we _could_ pull them and add them in the custom template...

**Comment by paaljoachim** [comment]
Next test. - Feedback from a call me user who has not played around a lot of FSE templates. -
I am brainstorming to what I see. I assume that a lot of my comments are far outside of this PR.
(This time I happen to use the TT1 theme.)

This area has been cleaned up:
Screen Shot 2021-04-06 at 12 48 28

What about loading in an existing custom template?
Will there be a collection of templates one can use?

Clicking New brings up the Name modal.
Screen Shot 2021-04-06 at 12 50 54

After saving the template name it goes into Template mode.
The one place I see it is under Status & visibility - Template. 

It looked like a prefix showed up: wp-custom-template and then the name I saved it as movie-page-template.

Screen Shot 2021-04-06 at 12 53 33

It would be helpful to add template info to the bottom Breadcrumbs + in the top bar. It is easy to overlook that one is in template mode as it looks almost identical to regular page content mode.

Ahh after clicking into the template layout the template name is suddenly seen in the top bar. (Or perhaps I had jumped around and that it would show up during edit template mode.)
One thing that gives a signal that it is in template mode is clicking the Post (Page) Title as it shows the title has become a block. 

It is a bit cramped in the top.
Screen Shot 2021-04-06 at 13 01 23

With the Top Toolbar active.
Screen Shot 2021-04-06 at 13 01 41

One can now cancel out of the template screen by clicking Cancel. (Thank you..:)

I went to another page and tested to see if the old template would show up. It did not. I made a new template. Then began modifying the page/post title. It looked like this.

Screen Shot 2021-04-06 at 13 32 06

I clicked Apply. This is seen.

Screen Shot 2021-04-06 at 13 34 47

I find it strange that the Page Save would show up in the Template mode. As only the Template save should be done here before going back to the Page content. Perhaps I need to do a few hard refreshes, as it should not be there. 
If I cancel out of the Save and go back to the Page content the dot in the Publish button is seen.
Again I assume I need to do some hard refreshing of the browser.

1. Go into Template mode.
2. Add some info.
3. Click to Apply.
4. Click Cancel.
5. It goes back to Page content but there is a dot in the Publish button.
6. Click Publish. Uncheck the template. 
7. Click Save.
8. Click the Publish button and notice it is not clickable.

This is the tricky part with the various dot saves I have seen multiple places. In FSE and Reusable blocks. 
One can get stuck somewhere in the Save -> Publish process. Forcing the user to Save any changes made.

**Comment by youknowriad** [comment]
> I find it strange that the Page Save would show up in the Template mode. 

In template mode, you don't edit just the template, you can edit everything, so the save flow shows all the unsaved changes.

**Comment by jameskoster** [comment]
> What should this header template contain? There's no way to know what the current theme contains.

I don't think it really matters. It's more about introducing the concept of a header template part, and circumventing scenarios where a user might recreate the same "header" multiple times for different custom templates.

It could be as simple as containing the Site Title and Site Tagline, centrally aligned.

**Comment by youknowriad** [comment]
I think ideally, it's the theme's decision to define the default template to be used in that case. 
I wouldn't mind a template using site title, site tagline and things like that but I don't think we should use template part personally to start with. I prefer for it to be minimal, we could ship this without shipping template parts.

**Comment by carolinan** [comment]
If there is a theme.json file with template parts assigned to template areas, I think those header and footer template parts should be included, but it is too early to guess if these themes will include a theme.json or not.

I think the UX needs to be improved and that the option needs to be integrated somehow with the existing "page attribute" where you select templates, but this would work for a POC and first version.

**Comment by youknowriad** [comment]
I've added an e2e test and a paragraph to docs to opt-out of this and also added some site blocks (site title, logo, tagline) to the default template.

I agree that the flows and design need to be iterated on.

What do you think. Ship it or not?

**Comment by aristath** [comment]
> What do you think. Ship it or not?

Yes!!

**Comment by youknowriad** [comment]
Then click the green button :P

**Review by aristath — APPROVED** [review]

**Comment by paaljoachim** [comment]
From my perspective this feature is a good one to have. Merging a bare bones version seems like a good idea. Of course there are a lot of followups based on the tests that I shared that needs to be done, but that is something one gradually can get to through various PR's. Thanks for creating the PR, Riad!

File: lib/init.php [hunk 2]
```diff
@@ -177,3 +177,4 @@ function register_site_icon_url( $response ) {
 add_filter( 'rest_index', 'register_site_icon_url' );
 
 add_theme_support( 'widgets-block-editor' );
+add_theme_support( 'block-templates' );
```
Comment (vindl) [inline lib/init.php:180]: @youknowriad I find this part a bit unexpected. Why are we unconditionally declaring theme support in the plugin/core? 🤔 

If we want to have this active by default, we could remove the theme support part altogether, and modify the activation checks to use the filter for opt-out, instead of:

> Theme authors can opt-out of this feature by removing the `block-templates` theme support in their `functions.php` file.

Comment (youknowriad) [reply lib/init.php:180]: This is something I explain here https://github.com/WordPress/gutenberg/pull/30465#discussion_r606174487

Instead of having a "disable-something" theme support, Core prefer to have theme supports enabled by default and plugins/themes can do `remove_theme_support` as it reads better. Adding a theme support calls "disable-something" doesn't read very well.

Comment (vindl) [reply lib/init.php:180]: Thanks for clarifying! While reading the activation conditions I assumed that this will be only enabled for themes that explicitly declare support, so I was a bit surprised afterward when I found out in testing that it actually works for all themes.

Comment (youknowriad) [reply lib/init.php:180]: yes, I the moment, this is opt-out because we want people to experience this, depending how the betas of WP go though, we might reconsider.

Comment (mtias) [reply lib/init.php:180]: To close the loop, #32858 makes it opt-in on classic themes and opt-out for themes using theme.json.
